### PR TITLE
Fixes #31745 - Add GraphQL client

### PR DESCRIPTION
**Pull request by xprazak2**
Adds Apollo client that allows communication
with GraphQL API.

The resurrection of https://github.com/theforeman/foreman/pull/6806

New dependencies in [foreman-js](https://github.com/theforeman/foreman-js/pull/256)

### CODE REVIEW

File: webpack/stories/docs/fetching-data-with-graphql.stories.mdx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+# Fetching data with GraphQL
+
+Foreman uses Apollo client to talk to GraphQL endpoint. All the components within react routes have access to the Apollo client as routes are wrapped with ApolloProvider (see ReactApp component for details). Queries can be executed simply by using a hook:
+
+```js
+import { useQuery, gql } from '@apollo/client';
+
+const myQuery = gql`
+  query {
+    whatever {
```
Comment (ofedoren) [inline webpack/stories/docs/fetching-data-with-graphql.stories.mdx]: Just a suggestion: wouldn't it be better to have an actual example? And add a comment that it should support whatever resource is currently supported by GraphQL, just to be over explicit...

Comment (xprazak2) [reply webpack/stories/docs/fetching-data-with-graphql.stories.mdx]: Thanks, I made an update as suggested.

**Comment by xprazak2** [comment]
Hmm, https://github.com/theforeman/foreman-js/pull/261 should fix the linting.

**Comment by jeremylenz** [comment]
@xprazak2 Keep in mind that before theforeman/foreman-js#261 can fix the linting, Foreman will need to consume version >= 8.3.3 of `@theforeman/vendor`.  To work around it until then you can `import { foo } from '@theforeman/vendor'`, but it would have to be changed back later.

**Comment by xprazak2** [comment]
Thanks @jeremylenz!

I opened https://github.com/theforeman/foreman/pull/8328 and [packaging update](https://github.com/theforeman/foreman-packaging/pull/6319) so that we have a proper fix and do not have to make changes later.

**Comment by xprazak2** [comment]
I needed to bump `eslint-plugin-foreman` and `vendor-dev` so that newer `vendor-core` is pulled in and linting finally passes.

**Comment by xprazak2** [comment]
The reason for changes in DiffView is https://github.com/ecomfe/gitdiff-parser/issues/12

**Comment by xprazak2** [comment]
All green now :)

File: webpack/assets/javascripts/react_app/Root/apollo.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { ApolloClient, ApolloLink, InMemoryCache, from } from '@apollo/client';
+import { BatchHttpLink } from '@apollo/client/link/batch-http';
+
+const batchLink = new BatchHttpLink({ uri: '/api/graphql' });
```
Comment (ofedoren) [inline webpack/assets/javascripts/react_app/Root/apollo.js]: Just to be sure, don't we need to use `foremanUrl` wrapper for every URL?

**Review by ofedoren — DISMISSED** [review]
Thanks, @xprazak2! LGTM, and I've tested a simple query in plugin: seems to be working just fine.

File: webpack/assets/javascripts/react_app/Root/apollo.js (hunk 2 above)
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/Root/apollo.js]: I am not sure either, seems like a we use it in other places so I added it here as well.

**Comment by xprazak2** [comment]
Are there any additional concerns? Updates are already merged in @theforeman/packaging

**Comment by ofedoren** [comment]
Not from me :) Although I wanted to ask, does it work for mutations as well?

**Comment by xprazak2** [comment]
Yes, there is [useMutation hook](https://www.apollographql.com/docs/react/data/mutations/#executing-a-mutation). And server has to know how to handle that particular mutation, of course.

**Review by kamils-iRonin — DISMISSED** [review]

**Review by tbrisker — APPROVED** [review]
Packaged foreman-vendor is already 8.3.3, only dependency changes left aren't packaged. packaging-wise 👍🏼

File: webpack/assets/javascripts/react_app/Root/apollo.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { ApolloClient, ApolloLink, InMemoryCache, from } from '@apollo/client';
+import { BatchHttpLink } from '@apollo/client/link/batch-http';
+
+import { foremanUrl } from '../common/helpers';
+
+const batchLink = new BatchHttpLink({ uri: foremanUrl('/api/graphql') });
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/Root/apollo.js:6]: Please add `graphql` to the eslint spell ignore

**Review by tbrisker — CHANGES_REQUESTED** [review]
Can you split out the diff view changes to a separate commit? just so it's clear what changes are actually related to this issue and what are just due to the dependency bump. Also add `graphql` to the ignored list so spelling lint doesn't warn about it.

**Comment by xprazak2** [comment]
I split the commits and made an update for spell checker.

**Review by tbrisker — APPROVED** [review]
Thanks @xprazak2 and all reviewers!

**Comment by ekohl** [comment]
Next time please increase the minimum version of the vendor package as well: https://github.com/theforeman/foreman/pull/8387

**Pull request merged**